src/validate.ts

```diff
@@ -26,18 +40,77 @@ function responseValidation(apiResponse: object, schema: object) {
 
 /**
  * Identifies the response type and validates the function based on the schemas.
- * @param apiResponse User defined api response
- * @param responseType User defined intent response
+ * @param req Request from one of the generate request functions.
+ * @param apiResponse User defined api response.
  * @return Errors from AJV validation, if any. Undefined otherwise.
  */
-export function validate(intentRequest: object, apiResponse: object) {
+export function validate(intentRequest: object, apiResponse: object, executeCommand?: string) {
   const responseType = intentRequest['inputs'][0]['intent'];
   if (responseType === 'action.devices.SYNC') {
-    return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+    // validate traits with attributes schema
+    const syncErrors : object[] = [];
+    const syncDevices = apiResponse['payload']['devices'];
+    const syncDevicesLength = syncDevices.length;
+    for (let i = 0; i < syncDevicesLength; i++) {
+      const traits = syncDevices[i]['traits'];
+      const traitsLength = syncDevices[i]['traits'].length;
+      const attributes = syncDevices[i]['attributes'];
+      if (attributes === undefined) {
+        return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+      }
+      for (let j = 0; j < traitsLength; j++) {
+        const trait = traits[j];
+        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
+          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+          if (validateTraitRes) {
+            return syncErrors.push(...validateTraitRes);
+          }
+        }
+      }
+    }
+    return syncErrors.length ? syncErrors : undefined;
   } else if (responseType === 'action.devices.QUERY') {
+    // validate with states schema;
+    const queryErrors : object[] = [];
+    const devices = intentRequest['inputs'][0]['payload']['devices'];
+    const devicesLength = devices.length;
+    for (let i = 0; i < devicesLength; i++) {
+      const deviceIds = devices[i]['id'];
+      const states = apiResponse['payload']['devices'][deviceIds];
+      if (executeCommand != undefined){
+          const validateQueryTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[executeCommand]);
```

Comment:
Should the execute command be the right input, or use the initial SYNC data to identify the device ID and all of its associated traits?

Reply:
Had a little trouble with using the SYNC data at first, so used execute command to pivot, but I think I identified a better way to pass in SYNC data now